### DISCUSSION TITLE
WebGL flickering on Samsung Galaxy S21 Android 12

### DISCUSSION THREAD

**Discussion by julien-moreau** [pixijs/pixijs#8038] [Q&A]
Hi all! I hope you are fine! :)

**Note:** I'm just reposting my topic here as asked by @ivanpopelyshev that was posted here: https://www.html5gamedevs.com/topic/47939-webgl-flickering-on-samsung-s21-android-12/ :)

We are experiencing a strange graphics issue using the Samsung s21 with Android 12 (even chrome, firefox, etc.). We have several games developed using the awesome Pixi.JS v4.8.8 and, at some particular state of the renderer, the screen displays some gliches/flickers each frame.

I'm part of the Babylon.JS core team so, for sure, I have firstly tried to reproduce using Babylon.JS :)
Unfortunately I failed to reproduce so I decided to dig into the Pixi.JS source code.

After some intuition and tries, I found that the problem is solved when all vertex buffers stored using the "STREAM_DRAW" hint become stored using the "STATIC_DRAW" hint. Both "STREAM_DRAW" and "DYNAMIC_DRAW" generate the problem, only "STATIC_DRAW" fixes the problem. Of course, I don't think the problem comes from Pixi.JS but more from the driver or the browser itself. It would be awesome if we can be sure of that together. As I'm not a Pixi.JS source code professional, I would really appreciate if you can confirm it can't come from the Pixi.JS WebGL renderer ❤️

I have shared 2 videos where you can see the flickering effect. You can also see that according to what is rendered, the problem looks different.
In the stage there are only sprites and only 1 mesh (NineSlicePlane), no filter and not render target. WebGL context is created using default values (stencil: true, alpha: false, etc.).

Also, I failed to reproduce in the Pixi.JS playground because I think the problem appears only in specific states, specific buffer sizes etc. Maybe that's why I failed to reproduce with Babylon.JS as well :(
When I say "specific", I mean that we have a lot of other games using Pixi.JS 4.8.8 and they are working perfectly with Samsung S21 & Android 12. Only few have this problem I mentioned above.

If needed, I can share some Spector.JS records so you can see the delta in WebGL commands between 2 frames (one that works and one that fails)

Here all the informations I have:

- the problem does not appear in Samsung S21 with Android 11
- the problem appears with Samsung S21 with Android 12
- the problem appears with Firefox on Samsung S21 with Android 12
- the problem does not appear in Google Pixel 5 with Android 12

Cheers and thanks again for your awesome work!

https://user-images.githubusercontent.com/2620127/145607492-44a5dcae-3fab-4c97-81b9-2931e3fd30c2.mov

https://user-images.githubusercontent.com/2620127/145607505-e486d999-b042-4404-a945-63ced5805c72.mov

**Comment by dijitaletkinlikler** [comment]
Hi. I think this problem is related to android webview. Android 6 has similar flickering issues.

**Reply by julien-moreau** [reply]
Interesting!
The problem here is that Firefox on Samsung S21 with Android 12 produces the exact same flickering issue. Is Firefox using Android/Chrome webviews or is it using their own?

**Reply by dijitaletkinlikler** [reply]
My phone is samsung j7 prime android 6 is installed.
When I disable Webview, the flickering goes away. When I activate Webview again, the same problem occurs. I don't know what Firefox is using.

**Comment by Excaa** [comment]
With pixi 4.8.9 changing STREAM_DRAW to STATIC_DRAW fixed the issue on S21. Don't have other Android 12 devices in use to confirm. Gonna test with other projects (starting from pixi 3.0.11 up to 4.8.9) later.

**Reply by julien-moreau** [reply]
@Excaa would be great if you can confirm :)
Here is informations I have:

- the problem does not appear in Samsung S21 with Android 11
- the problem appears with Samsung S21 with Android 12
- the problem appears with Firefox on Samsung S21 with Android 12
- the problem does not appear in Google Pixel 5 with Android 12

**Reply by Excaa** [reply]
Only one S21 device current and that's already updated to Android 12. Though our testers didnt report anything when it was 11 so most likely problem didnt exist at that point.

We see the problem with 
- Samsung S21 Android 12 Chrome  96.0.4664.45
- Samsung S21 Android 12 SamsunbBrowser 16.0.2.19

No other Android 12 devices yet.

**Reply by Excaa** [reply]
Changed all STREAM_DRAWS to STATIC_DRAW in pixi 4.x and DYNAMIC_DRAW to STATIC_DRAW in 3.x and all projects have worked without a problem (only Chrome tested so far).

**Reply by sam-derivco** [reply]
Hi there. Any updates on this issue? Would you be able to share any console logs you can gather? And maybe a SpectorJS profile? I'm also seeing similar issues on the Samsung S21 with Android 12.

**Reply by julien-moreau** [reply]
Hello @sam-derivco !
If you reproduce you can try to create a SpectorJS record yourself. Something interested would be to try catching a working frame and a glitched frame to see if there's any difference (i doubt but to be sure). For instance, I failed to record a glitched frame but more we are, more we can :)

**Reply by sam-derivco** [reply]
@julien-moreau I would love to, but the only device I have access to is through a remote connection. :( And I'm not sure if it's possible to use SpectorJS in that way... Is it?

**Reply by Excaa** [reply]
Our QA got Nokia x20 with Android 12 and that didn't have the issue when testing without changing draws. Need to check more devices with Android 12 when they roll out to see if only S21 is suffering from the issue.  

Havent got access to S21 so cant get more exact data.

**Reply by sam-derivco** [reply]
@Excaa Yeah I think this is some sort of a Samsung driver problem. I've also noticed that the S21 and S21 Ultra have 128 fragment texture units for WebGL 1.0. This is very strange, since most modern devices only have 16. (It actually caused some additional problems on my side). Coincidentally, these are the only devices we can replicate this flickering problem on. The S21+ is fine.

**Comment by dijitaletkinlikler** [comment]
My phone is samsung j7 prime android 6 is installed.
When I disable Webview, the flickering goes away. When I activate Webview again, the same problem occurs. I don't know what Firefox is using.

This situation is not related to pixijs in my opinion. my apps built with both pixijs and createjs are all flickering. But I can only test with android 6.

**Reply by Excaa** [reply]
Have you tried if changing STREAM_DRAW / DYNAMIC_DRAW to STATIC_DRAW helps with that? Though I suspect it's a different issue since rendering between Android 6 and 12 differ quite much.

**Reply by dijitaletkinlikler** [reply]
No, I haven't tried. I don't know how to use this. If you write the web address of the example you made, I can try it on my phone and get back to you.

**Reply by Excaa** [reply]
Just replace STATIC_DRAW & DYNAMIC_DRAW in pixi.js file to STATIC_DRAW. Havent got a public example for that as I have same issue as julien-moreau that simple examples don't have the issue.

**Comment by tmarkovic** [comment]
One of our QAs got build G991BXXU4BULF/G991BOXM4BULF/G991BXXU4BULG on their Galaxy s21 5G (SM-G991B*) and the previous flickering we experienced can no longer be reproduced!

**Reply by macguffin** [reply]
Hi @tmarkovic do you know if this build was an automatic update? Did the device previously have the affected A12 OS or was it a new update from A11?

Thanks for the info :)

**Reply by tmarkovic** [reply]
Hmm I'm a bit unsure as to what you mean about automatic but it was OTA so to speak (not installed manually, provided by manufacturer). There was no bump in either Android or One UI version (12 and 4.0).

The device could reproduce the rendering issues previous to the update.

No problem, such a relief 👍 

I'll keep you posted if anything changes!

**Reply by iqbalsiraj** [reply]
@tmarkovic Was the device a UK or International model? (UK has Exynos and International has Snapdragon processor)

**Reply by tmarkovic** [reply]
> @tmarkovic Was the device a UK or International model? (UK has Exynos and International has Snapdragon processor)

Device was bought in Sweden!

**Reply by julien-moreau** [reply]
Hi @tmarkovic, @iqbalsiraj I can validate that my French device has been updated yesterday and no more flickering issue ! Looks like it was finally a Samsung or Android issue. Have you tried to update your phone since 15 days?

**Reply by iqbalsiraj** [reply]
> Hi @tmarkovic, @iqbalsiraj I can validate that my French device has been updated yesterday and more flickering issue ! Looks like it was finally a Samsung or Android issue. Have you tried to update your phone since 15 days?

@julien-moreau Did you update previously and it fixed and now you saying there's another update and it has broken it again?

**Reply by julien-moreau** [reply]
Sorry @iqbalsiraj for late and I just edited my post! I wanted to say « no more issues » sorry for ambiguity!

For my part I can’t reproduce, it looks like it has been fixed :)

**Comment by kenrussell** [comment]
Can any test case that reproduces this problem be provided? Without one it'll be difficult or impossible to guess what the problem was - and make sure it is fixed on all similar devices. Thanks in advance.

**Reply by surtr-isaz** [reply]
I've managed to narrow this issue down. It occurs when uploading to the same buffer in the same frame but only when greater than a specific size - I believe there is an optimised pathway for smaller buffers hence why this is tricky to reproduce in all applications.

I know this is a bit vague, but I intend on putting together a raw WebGL demo this week and submitting it to Google as a bug report. I'll drop a link here when I do.

FWIW: We've added means to circumvent this issue on specific devices in on our internal renderer, but as a workaround in Pixi developers can set the **`PIXI.settings.CAN_UPLOAD_SAME_BUFFER`** flag to **`false`**.

In my report I'll also include a list of all known affected devices (at least to me) which is a list of about 38 model numbers/codes.

**Reply by surtr-isaz** [reply]
Also, some extra stuffs, while this seems to be resolved on the S21 family of devices with the OneUI update, it is still present on the Pixel 6 family.

Also also, while some devs were forcing STATIC_DRAW to get around the problem on the S21 family this did not resolve the issue on the Pixel 6 family. The workaround I suggested works on both families of devices.

**Reply by martinnikolov93** [reply]
I've set up a codepen [here](https://codepen.io/martin-nikolov/pen/xxpyOyz)

On pixel 6 the flickering occurs with at least 65 sprites and 1 sprite with a mask. Setting `CAN_UPLOAD_SAME_BUFFER` to `false` stops flickering as @surtr-isaz suggested.

**Reply by surtr-isaz** [reply]
Awesome, thanks @martinnikolov93!

I still haven't gotten around to making that demo. However, I recall reading something somewhere that there is an optimization in the new Vulkan backend for smaller buffers. This pricked up my ears because I was only seeing this issue occur when batching geometry in excess of a specific size (can't recall right now, need to consult old notes, but it wasn't much) and uploading via bufferSubData where it was likely falling back to the original path(?).

Anyway, if that all sounds a bit vague it's because it is :D I haven't had a chance to dig deeper into it ...

**Reply by kenrussell** [reply]
Thanks for the demo! We can confirm that a couple of colleagues can reproduce the flickering on Pixel 6.

@surtr-isaz which Vulkan backend were you referring to?

We want to make it possible to try ANGLE's new Vulkan backend in Chrome as a potential mitigation for this issue, which might be a bug in the OpenGL ES driver. Right now this is not feasible for most users. I've filed http://crbug.com/1317775 regarding adding this setting to Chrome.

**Comment by liwaprince** [comment]
thats the same graphics issue i get on Dolphin emulator when i run it with OpenGl mode.

i switched to Vulkan mode, it fixes the problem but its way slower.

Galaxy s21 Ultra Exynos, this issue started after updating to Android 12.

**Comment by surtr-isaz** [comment]
I believe I've found the cause for this bug and it exists in the OS/browser. It's related to uploading to same buffer in the same frame but only when greater than a specific size. There might be an optimised pathway for smaller buffers hence why this is tricky to reproduce in all applications.

We've added means to circumvent this issue on specific devices in on our internal renderer, but as a workaround for Pixi developers you can set the **`PIXI.settings.CAN_UPLOAD_SAME_BUFFER`** flag to **`false`**.

**Reply by surtr-isaz** [reply]
Also, see my comments above to @kenrussell for some extra information.

**Reply by djlastnight** [reply]
`PIXI.settings.CAN_UPLOAD_SAME_BUFFER = false;` works like a charm for pixi 5.3.11 on S22 android 12 chrome. Thank you!
